lib/control-connection.js

```diff
@@ -378,38 +294,133 @@ ControlConnection.prototype.refreshHosts = function (newNodesUp, callback) {
   ], callback);
 };
 
-ControlConnection.prototype.hostDownHandler = function () {
-  this.log('warning', 'Host ' + this.host.address + ' used by the ControlConnection DOWN');
-  this.host = null;
-  this.connection = null;
-  this.refresh();
-};
-
 /**
- * Acquires a connection and refreshes topology metadata.
+ * Acquires a connection and refreshes topology and keyspace metadata.
+ * <p>If it fails obtaining a connection:</p>
+ * <ul>
+ *   <li>
+ *     When its initializing, it should:
+ *     <ul>
+ *       <li>Continue iterating through the hosts</li>
+ *       <li>When there aren't any more hosts, it should invoke callback with the inner errors</li>
+ *     </ul>
+ *   </li>
+ *   <li>
+ *     When its running in the background, it should:
+ *     <ul>
+ *       <li>Continue iterating through the hosts</li>
+ *       <li>
+ *         When there aren't any more hosts, it should:
+ *         <ul>
+ *           <li>Schedule reconnection</li>
+ *           <li>Invoke callback with the inner errors</li>
+ *         </ul>
+ *       </li>
+ *     </ul>
+ *   </li>
+ * </ul>
+ * <p>If it fails obtaining the metadata, it should:</p>
+ * <ul>
+ *   <li>It should mark connection and/or host unusable</li>
+ *   <li>Retry using the same iterator from query plan / host list</li>
+ * </ul>
+ * @param {Boolean} [reuseQueryPlan]
  * @param {Function} [callback]
  */
-ControlConnection.prototype.refresh = function (callback) {
+ControlConnection.prototype.refresh = function (reuseQueryPlan, callback) {
+  var initializing = !this.initialized;
+  callback = callback || utils.noop;
+  // Reset the state of the host field, that way we can identify when the query plan was exhausted
+  this.host = null;
   var self = this;
   utils.series([
-    this.getConnection.bind(this),
-    function (next) {
-      if (!self.connection) {
+    function getHostIterator(next) {
+      if (reuseQueryPlan) {
         return next();
       }
-      self.refreshOnConnection(false, next);
-    }], function doneRefreshing(err) {
-
-    if (err || !self.connection) {
-      self.log('error', 'ControlConnection was not able to reconnect');
+      self.triedHosts = {};
+      if (initializing) {
+        self.log('info', 'Getting first connection');
+        self.hostIterator = utils.arrayIterator(self.hosts.values());
+        return next();
+      }
+      self.log('info', 'Trying to acquire a connection to a new host');
+      self.profileManager.getDefaultLoadBalancing().newQueryPlan(null, null, function onNewPlan(err, iterator) {
+        if (err) {
+          self.log('error', 'ControlConnection could not retrieve a query plan to determine which hosts to use', err);
+          return next(err);
+        }
+        self.hostIterator = iterator;
+        next();
+      });
+    },
+    function getConnectionTask(next) {
+      self.borrowAConnection(next);
```

Comment:
I see, I think the act of calling `borrowConnection` on a Host with an empty pool will create a connection.